symbolic-debuginfo/src/elf.rs

```diff
@@ -550,37 +557,60 @@ impl<'data, 'object> Iterator for ElfSymbolIterator<'data, 'object> {
-        for symbol in &mut self.symbols {
-            // Only check for function symbols.
-            if symbol.st_type() != elf::sym::STT_FUNC {
-                continue;
-            }
+        fn get_symbols<'data>(
+            symbols: &mut SymIterator,
+            strtab: &Strtab<'data>,
+            load_addr: u64,
+            sections: &[SectionHeader],
+        ) -> Option<Symbol<'data>> {
+            for symbol in symbols {
+                // Only check for function symbols.
+                if symbol.st_type() != elf::sym::STT_FUNC {
+                    continue;
+                }
 
-            // Sanity check of the symbol address. Since we only intend to iterate over function
-            // symbols, they need to be mapped after the image's load address.
-            if symbol.st_value < self.load_addr {
-                continue;
-            }
+                // Sanity check of the symbol address. Since we only intend to iterate over function
+                // symbols, they need to be mapped after the image's load address.
+                if symbol.st_value < load_addr {
+                    continue;
+                }
+
+                let section = match symbol.st_shndx {
+                    self::SHN_UNDEF => None,
+                    index => sections.get(index),
+                };
 
-            let section = match symbol.st_shndx {
-                self::SHN_UNDEF => None,
-                index => self.sections.get(index),
-            };
+                // We are only interested in symbols pointing into sections with executable flag.
+                if !section.map_or(false, |header| header.is_executable()) {
+                    continue;
+                }
 
-            // We are only interested in symbols pointing into sections with executable flag.
-            if !section.map_or(false, |header| header.is_executable()) {
-                continue;
-            }
+                let name = strtab.get_at(symbol.st_name).map(Cow::Borrowed);
 
-            let name = self.strtab.get_at(symbol.st_name).map(Cow::Borrowed);
+                return Some(Symbol {
+                    name,
+                    address: symbol.st_value - load_addr,
+                    size: symbol.st_size,
+                });
+            }
 
-            return Some(Symbol {
-                name,
-                address: symbol.st_value - self.load_addr,
-                size: symbol.st_size,
-            });
+            None
         }
 
-        None
+        if let Some(debug_symbol) = get_symbols(
+            &mut self.symbols,
+            self.strtab,
+            self.load_addr,
+            self.sections,
+        ) {
+            Some(debug_symbol)
+        } else {
+            get_symbols(
+                &mut self.dynamic_symbols,
+                self.dynamic_strtab,
+                self.load_addr,
+                self.sections,
+            )
+        }
```

Comment:
Personally, I think this reads better with `or_else`:
```suggestion
        get_symbols(
            &mut self.symbols,
            self.strtab, 
            self.load_addr,
            self.sections,
        )
        .or_else(|| {
            get_symbols(
                &mut self.dynamic_symbols,
                self.dynamic_strtab,
                self.load_addr,
                self.sections,
            )
        })
```

Reply:
Yes, much better. I'll put this change into the patch and push again. I also considered chaining `self.elf.syms` and `self.elf.dynsyms` when iterating over the symbols but then I couldn't find a simple way to select between `strtab` and `dynstrtab` so I went for this cruder approach.